avalon/nuke/lib.py

```diff
@@ -23,48 +24,255 @@ def maintained_selection():
+
+        elif isinstance(value, int):
+            knob = nuke.Int_Knob(name, nice)
+            knob.setValue(value)
+
+        elif isinstance(value, six.string_types):
+            knob = nuke.String_Knob(name, nice)
+            knob.setValue(value)
+
+        elif isinstance(value, list):
+            knob = nuke.Enumeration_Knob(name, nice, value)
+
+        elif isinstance(value, dict):
+            if all(isinstance(v, dict) for v in value.values()):
+                # Create a group of tabs
+                begain = nuke.BeginTabGroup_Knob()
+                end = nuke.EndTabGroup_Knob()
+                begain.setName(name)
+                end.setName(name + "_End")
+                knobs.append(begain)
+                for k, v in value.items():
+                    knobs += create_knobs(v, tab=k)
+                knobs.append(end)
+            else:
+                # Create a group of knobs
+                knobs.append(nuke.Tab_Knob(name, nice, nuke.TABBEGINGROUP))
+                knobs += create_knobs(value)
+                knobs.append(nuke.Tab_Knob(name, nice, nuke.TABENDGROUP))
+            continue
+
+        else:
+            raise TypeError("Unsupported type: %r" % type(value))
+
+        knobs.append(knob)
+
+    return knobs
 
 
 def add_publish_knob(node):
+    """Add Publish knob to node
+
+    Arguments:
+        node (obj): nuke node to be processed
+
+    Returns:
+        node (obj): processed nuke node
+
+    """
     if "publish" not in node.knobs():
-        knob = nuke.Boolean_Knob("publish", "Publish")
-        knob.setFlag(0x1000)
-        knob.setValue(False)
-        node.addKnob(knob)
+        body = OrderedDict()
+        body[("divd", "")] = Knobby("Text_Knob", "")
+        body["publish"] = True
+        imprint(node, body)
+    return node
+
+
+def set_avalon_knob_data(node, data=None, prefix="avalon:"):
+    """ Sets data into nodes's avalon knob
+
+    Arguments:
+        node (obj): Nuke node to imprint with data,
+        data ( dict): Data to be imprinted into AvalonTab
+        prefix (str, optional): filtering prefix
+
+    Returns:
+        node (obj)
+
+    Examples:
+        data = {
+            'asset': 'sq020sh0280',
+            'family': 'render',
+            'subset': 'subsetMain'
+        }
+    """
+    editable = ["asset", "subset", "name", "namespace"]
+
+    data = data or dict()
+
+    group = OrderedDict()
+    body = OrderedDict()
+
+    group["avalonDataGroup"] = body
+    body[("warn", "")] = Knobby("Text_Knob",
+                                "Warning! Do not change following data!")
+    body[("divd", "")] = Knobby("Text_Knob", "")
+
+    for key, value in data.items():
+        name = prefix + key
+        if key in editable:
+            body[name] = value
+        else:
+            body[name] = Knobby("Text_Knob", str(value))
+
+    imprint(node, group, tab="AvalonTab")
+
     return node
 
 
+def get_avalon_knob_data(node, prefix="avalon:"):
```

Comment:
Would it be possible to turn this into `prefix=["avalon:",  "ak:"]` for backward compatibility with already running productions? it was previously `prefix="ak:"`

Reply:
Ah, okay, but I won't put mutable value as arg's default value, I might hard code it in the for loop below :)

Speaking of which, was there any reason for applying prefix on knobs ?
Would it be okay if we not using prefix in future (with backward compatibility of course) ?

Reply:
> Would it be okay if we not using prefix in future (with backward compatibility of course) ?

If you found the efficient way to get efficient way of filtering only avalon attributes from all knobs then yes. The reason was to separate only avalon metadata attributes, but didm't see other way to do it ;)

Reply:
I have found a way to list out all user-defined knobs in any given node, just like other host implementations (`lib.read`). But noted that user-defined layout knobs like tab or group will be listed too.

Here's the gist for trying:
https://gist.github.com/davidlatwe/519528bb01b4ff816389464ec2709e97

Maybe we could work without prefixing knob name if this works ?
Anyway, I'll try resolving this in another PR, this PR will be merged as is for now. :relaxed:

Reply:
Interesting solution!
Listing user defined knobs would definitely be useful.

Reply:
@jezscha  I end up implementing the backward compatibility in a new method `lib.read`, instead of changing `lib.get_avalon_knob_data`. Please let me know how you think of this, it's in PR #476 :smiley: